File: src/#1030/solution.js
var allCellsDistOrder = function (R, C, r0, c0) {
  let visited = []
  for (let i = 0; i < R; i++) {
    visited[i] = new Array(C)
  }
  
  let directions = [
    [0, 1],
    [0, -1],
    [1, 0],
    [-1, 0],
  ]

  let queue = [[r0, c0]]
  visited[r0][c0] = true
  let ret = []

  while (queue.length !== 0) {
    let front = queue[0]
    ret.push(front)
    queue.shift()
    directions.forEach(d => {
      let x = front[0] + d[0]
      let y = front[1] + d[1]
      if (x >= 0 && y >= 0 && x < R && y < C && !visited[x][y]) {
        visited[x][y] = true
        queue.push([x, y])
      }
    })
  }

  return ret
}

File: src/#1030/solution.go
package solution

func allCellsDistOrder(R int, C int, r0 int, c0 int) [][]int {
	visited := make([][]bool, R)
	for i := range visited {
		visited[i] = make([]bool, C)
	}

	directions := [4][2]int{{0, 1}, {0, -1}, {1, 0}, {-1, 0}}

	queue := [][]int{{r0, c0}}
	visited[r0][c0] = true
	ret := make([][]int, 0, R*C)

	for len(queue) != 0 {
		front := queue[0]
		ret = append(ret, front)
		queue = queue[1:]
		for _, d := range directions {
			if x, y := front[0]+d[0], front[1]+d[1]; x >= 0 && y >= 0 && x < R && y < C && !visited[x][y] {
				visited[x][y] = true
				queue = append(queue, []int{x, y})
			}
		}
	}
	return ret
}
